framework/src/modules/interoperability/mainchain/module.ts

```diff
@@ -419,4 +438,108 @@ export class MainchainInteroperabilityModule extends BaseInteroperabilityModule
 			throw new Error('Invalid validatorsHash from chainData.lastCertificate.');
 		}
 	}
+
+	// https://github.com/LiskHQ/lips/blob/main/proposals/lip-0045.md#mainchain
+	private _verifyTerminatedStateAccounts(
+		chainInfos: ChainInfo[],
+		terminatedStateAccounts: TerminatedStateAccountWithChainID[],
+	) {
+		for (const chainInfo of chainInfos) {
+			const { chainID } = chainInfo;
+
+			const filteredTerminatedStateAccounts = terminatedStateAccounts.filter(a =>
+				a.chainID.equals(chainID),
+			);
+			if (terminatedStateAccounts.length > 0 && filteredTerminatedStateAccounts.length === 0) {
+				throw new Error('there can not be a terminated account if there is no chain account');
+			}
+			if (filteredTerminatedStateAccounts.length > 0) {
+				// For each entry chainInfo in chainInfos, chainInfo.chainData.status == CHAIN_STATUS_TERMINATED
+				// if and only if a corresponding entry (i.e., with chainID == chainInfo.chainID) exists in terminatedStateAccounts.
+				if (chainInfo.chainData.status !== ChainStatus.TERMINATED) {
+					throw new Error(
+						`chainInfo.chainData.status must be ${ChainStatus.TERMINATED} if chainInfo.chainID exists in terminatedStateAccounts.`,
+					);
+				}
+			}
+		}
+
+		// Each entry stateAccount in terminatedStateAccounts has a unique stateAccount.chainID
+		const chainIDs = terminatedStateAccounts.map(a => a.chainID);
+		if (!bufferArrayUniqueItems(chainIDs)) {
+			throw new Error(`terminatedStateAccounts don't hold unique chainID.`);
+		}
+
+		// terminatedStateAccounts is ordered lexicographically by stateAccount.chainID
+		const sortedByChainID = [...terminatedStateAccounts].sort((a, b) =>
+			a.chainID.compare(b.chainID),
+		);
+		for (let i = 0; i < terminatedStateAccounts.length; i += 1) {
+			if (!terminatedStateAccounts[i].chainID.equals(sortedByChainID[i].chainID)) {
+				throw new Error('terminatedStateAccounts must be ordered lexicographically by chainID.');
+			}
+		}
+
+		// For each entry stateAccount in terminatedStateAccounts holds
+		// stateAccount.stateRoot == chainData.lastCertificate.stateRoot,
+		// stateAccount.mainchainStateRoot == EMPTY_HASH, and
+		// stateAccount.initialized == True.
+		// Here chainData is the corresponding entry (i.e., with chainID == stateAccount.chainID) in chainInfos.
+		for (const chainInfo of chainInfos) {
```

Comment:
can this loop be merged with the above loop instead of looping twice over the chainInfos? after we check the stateAccount exists for each chainData.status=terminated we can have next check to see of stateRoot is correct